libraries/AP_HAL_ChibiOS/GPIO.cpp

```diff
@@ -300,6 +297,11 @@ AP_HAL::DigitalSource* GPIO::channel(uint16_t pin)
 
 extern const AP_HAL::HAL& hal;
 
+#if !defined(HAL_BOOTLOADER_BUILD)
```

Comment:
I really don't like having "what is my current target?" spread through the code any more than we need it to - and we don't need it to here.

A new define `HAL_CHIBIOS_PAL_INTERRUPT_ENABLED` would be nice.  `chibios_hwdef.py` can set it to zero for a bootloader build and it can be defaulted in `chibios.h`.

Might let us save a few bytes on some peripherals, too.

Reply:
A little bit confused why you propose making such a define ChibiOS-specific.

The real meat of this change is `#define PAL_USE_CALLBACKS FALSE` which instructs ChibiOS to not include the interrupt handlers. But then the relevant setup functions also go away and we have to remove these function implementations to avoid compile time errors.

So Proposal A: have a define for all HALs `HAL_SUPPORT_PIN_INTERRUPTS` which when off cuts out everything from all HALs. Then somewhere (probably `mcuconf.h`?) we have to make sure to `#define PAL_USE_CALLBACKS FALSE` when it's off.

Or Proposal B: just make this check use `PAL_USE_CALLBACKS` directly. We shouldn't need to make it conditional on the platform as the files shouldn't be run through the compiler on non-ChibiOS.

Reply:
> I'd be OK with that given we go onto use symbols almost certainly protected by that define...

Sounds good, we do which is why I had to put some sort of define here.